edge/pkg/devicetwin/dtmanager/communicate_test.go

```diff
@@ -0,0 +1,319 @@
+	"github.com/kubeedge/kubeedge/edge/pkg/devicetwin/dtcommon"
+	"github.com/kubeedge/kubeedge/edge/pkg/devicetwin/dtcontext"
+	"github.com/kubeedge/kubeedge/edge/pkg/devicetwin/dttype"
+)
+
+// TestStart is function to test Start().
+func TestStart(t *testing.T) {
+	mainContext := context.GetContext(context.MsgCtxTypeChannel)
+	dtContexts, _ := dtcontext.InitDTContext(mainContext)
+	heartChan := make(chan interface{}, 1)
+	receiveChan := make(chan interface{}, 1)
+	receiveChan <- &dttype.DTMessage{Action: "action", Identity: "identity", Msg: &model.Message{Content: "msg"}}
+	heartChan <- "stop"
+	tests := []struct {
+		name   string
+		Worker Worker
+		Group  string
+	}{
+		{
+			name: "StartTest",
+			Worker: Worker{
+				ReceiverChan:  receiveChan,
+				HeartBeatChan: heartChan,
+				DTContexts:    dtContexts,
+			},
+			Group: "",
+		},
+	}
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			cw := CommWorker{
+				Worker: test.Worker,
+				Group:  test.Group,
+			}
+			go cw.Start()
+		})
+	}
+}
+
+// TestDealSendToEdge is function to test dealsendToedge().
+func TestDealSendToEdge(t *testing.T) {
+	mainContext := context.GetContext(context.MsgCtxTypeChannel)
+	dtContexts, _ := dtcontext.InitDTContext(mainContext)
+	tests := []struct {
+		name     string
+		context  *dtcontext.DTContext
+		resource string
+		msg      interface{}
+		want     interface{}
+		wantErr  error
+	}{
+		{
+			name:     "dealSendToEdgeTest",
+			context:  dtContexts,
+			resource: "",
+			msg:      &model.Message{},
+			wantErr:  nil,
+			want:     nil,
+		},
+	}
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			got, err := dealSendToEdge(test.context, test.resource, test.msg)
+			if !reflect.DeepEqual(err, test.wantErr) {
+				t.Errorf("dealSendToEdge() error = %v, wantErr %v", err, test.wantErr)
+				return
+			}
+			if !reflect.DeepEqual(got, test.want) {
+				t.Errorf("dealSendToEdge() = %v, want %v", got, test.want)
+			}
+		})
+	}
+}
+
+// TestDealSendToCloud is function to test dealSendToCloud().
+func TestDealSendToCloud(t *testing.T) {
+	mainContext := context.GetContext(context.MsgCtxTypeChannel)
+	dtContextStateDisconnected, _ := dtcontext.InitDTContext(mainContext)
+	dtContextStateConnected, _ := dtcontext.InitDTContext(mainContext)
+	dtContextStateConnected.State = dtcommon.Connected
+	tests := []struct {
+		name     string
+		context  *dtcontext.DTContext
+		resource string
+		msg      interface{}
+		want     interface{}
+		wantErr  error
+	}{
+		{
+			name:     "dealSendToCloudTest-StateDisconnected",
+			context:  dtContextStateDisconnected,
+			resource: "",
+			msg:      "",
+			wantErr:  nil,
+			want:     nil,
+		},
+		{
+			name:     "dealSendToCloudTest-StateConnected",
+			context:  dtContextStateConnected,
+			resource: "",
+			msg:      "",
+			want:     nil,
+			wantErr:  errors.New("msg not Message type"),
+		},
+		{
+			name:     "dealSendToCloudTest-ActualMsg",
```

Comment:
In the case of a correct message being sent , the context's confirmMap stores the msg id along with the DTMessage...how can we test this ?

Reply:
Updated